client/consensus/src/collators.rs

```diff
@@ -0,0 +1,402 @@
+            None => return Ok(None),
+        }
+    };
+
+    Ok(Some(SlotClaim::unchecked::<P>(author_pub, slot)))
+}
+
+/// Attempt to claim a slot using a keystore.
+pub fn claim_slot_inner<P: Pair>(
+    slot: Slot,
+    authorities: &Vec<AuthorityId<P>>,
+    keystore: &KeystorePtr,
+    force_authoring: bool,
+) -> Option<P::Public>
+where
+    P: Pair,
+    P::Public: Codec + std::fmt::Debug,
+    P::Signature: Codec,
+{
+    let expected_author = crate::slot_author::<P>(slot, authorities.as_slice());
+    // if not running with force-authoring, just do the usual slot check
+    if !force_authoring {
+        expected_author.and_then(|p| {
+            if keystore.has_keys(&[(p.to_raw_vec(), NIMBUS_KEY_ID)]) {
+                Some(p.clone())
+            } else {
+                None
+            }
+        })
+    }
+    // if running with force-authoring, as long as you are in the authority set,
+    // propose
+    else {
+        authorities
+            .iter()
+            .find(|key| keystore.has_keys(&[(key.to_raw_vec(), NIMBUS_KEY_ID)]))
+            .cloned()
+    }
+}
+
+/// Seal a block with a signature in the header.
+pub fn seal_tanssi<B: BlockT, P>(
+    pre_sealed: B,
+    storage_changes: StorageChanges<HashingFor<B>>,
+    author_pub: &P::Public,
+    keystore: &KeystorePtr,
+) -> Result<BlockImportParams<B>, Box<dyn Error + Send + Sync + 'static>>
+where
+    P: Pair,
+    P::Signature: Codec + TryFrom<Vec<u8>>,
+    P::Public: AppPublic,
+{
+    let (pre_header, body) = pre_sealed.deconstruct();
+    let pre_hash = pre_header.hash();
+    let block_number = *pre_header.number();
+
+    // sign the pre-sealed hash of the block and then
+    // add it to a digest item.
+    let signature = Keystore::sign_with(
+        keystore,
+        <AuthorityId<P> as AppCrypto>::ID,
+        <AuthorityId<P> as AppCrypto>::CRYPTO_ID,
+        author_pub.as_slice(),
+        pre_hash.as_ref(),
+    )
+    .map_err(|e| sp_consensus::Error::CannotSign(format!("{}. Key: {:?}", e, author_pub)))?
+    .ok_or_else(|| {
+        sp_consensus::Error::CannotSign(format!(
+            "Could not find key in keystore. Key: {:?}",
+            author_pub
+        ))
+    })?;
+    let signature = signature
+        .clone()
+        .try_into()
+        .map_err(|_| sp_consensus::Error::InvalidSignature(signature, author_pub.to_raw_vec()))?;
+
+    let signature_digest_item = <DigestItem as NimbusCompatibleDigestItem>::nimbus_seal(signature);
+
+    // seal the block.
+    let block_import_params = {
+        let mut block_import_params = BlockImportParams::new(BlockOrigin::Own, pre_header);
+        block_import_params.post_digests.push(signature_digest_item);
+        block_import_params.body = Some(body.clone());
+        block_import_params.state_action =
+            StateAction::ApplyChanges(sc_consensus::StorageChanges::Changes(storage_changes));
+        block_import_params.fork_choice = Some(ForkChoiceStrategy::LongestChain);
```

Comment:
I am not sure about this one, specially looking at how polkadot-sdk hadnles this: https://github.com/paritytech/polkadot-sdk/blob/53f615de503c8936c2639f51b99019c64eab1018/cumulus/client/consensus/common/src/parachain_consensus.rs#L458

Reply:
Ah never mind, I see that they have longest chain too https://github.com/paritytech/polkadot-sdk/blob/53f615de503c8936c2639f51b99019c64eab1018/cumulus/client/consensus/aura/src/collator.rs#L373